### DISCUSSION TITLE
Generic Static Protocols Help

### DISCUSSION THREAD

**Discussion by InvincibleRMC** [python/typing#1854] [Q&A]
Since ClassVar's of Type Variable is not allowed is there anything better that can be done then `BadSrv`?

```python
from typing import ClassVar, TypeVar, Protocol

class Msg(Protocol): ...

RequestT = TypeVar('RequestT', bound=Msg)
ResponseT = TypeVar('ResponseT', bound=Msg)

#Ideal
class Srv(Protocol[RequestT, ResponseT]):
    Request: ClassVar[type[RequestT]]
    Response: ClassVar[type[ResponseT]]

#Jank with TypeAlias, but not supported in protocol
class SrvTypeAlias(Protocol[RequestT, ResponseT]):
    type Request = type(RequestT)
    type Response = type(ResponseT)
 
#Current Solution
 class BadSrv(Protocol):
    Request: ClassVar[type[Msg]]
    Response: ClassVar[type[Msg]]


#Example Service for reference
class DescribeParameters:
    from rcl_interfaces.srv._describe_parameters import DescribeParameters_Request as Request
    from rcl_interfaces.srv._describe_parameters import DescribeParameters_Response as Response
```

**Comment by erictraut** [comment]
Here are a couple of suggestions.

### Option 1: Don't use a `ClassVar`

Simply omit the `ClassVar` in the protocol definition. Type checkers will then treat these attributes as instance variables. The only downside to this approach is that users of the protocol will be able to overwrite the default value of the attribute (which is a class variable) with an instance value. That's probably something you'd want a type checker to prevent, but there's currently no official way to do this in the typing spec.

This limitation was discussed in the typing forums a while back. In response, I added support in pyright for `Final` attributes in a protocol definition. The typing spec doesn't officially allow or disallow `Final` attributes in a protocol. It's something that we could formalize if it's generally useful. Mypy currently rejects this usage.

Here's what that would look like...

Code sample in [pyright playground](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoBiqAhgDYA0UACuDGAMZikCwAUG-acQM7dQCy3NAAoaYOo1IBKAFxQAdIrYcuvKAGUQANwDaAJQCmARwCuB7jAAqcwWkqHuOFNwPWBQgLqjaDJrLZQgVCGpuYwckQoZDrwCAb6xmYWlh4eAUEOTi4RJKQxiPGZYM6uqcqsnDx8IUkwAILCtrIKShWq1eZZBg1NcoryKlUaBtpI9N0y6YE1YTlRpFAAvMGJYXVTK47F2YS5S5td6%2BysACYGwFBoBjAA%2BiCrFned24UPVjZC9s8lbrZeGwAPOSaXQzZJfLY-DzkNhSKAAWgAfLATAhSK9QuCDi8Un1WmwtPsrrd7piSd8XMJ1CMtGNulI2PctAYyDdYgZhFo4VAAMQotEY2p1CGHDxAA)

```python
from typing import Final, Protocol

class Msg(Protocol): ...

class Srv[RequestT: Msg, ResponseT: Msg](Protocol):
    Request: Final[type[RequestT]]
    Response: Final[type[ResponseT]]

class RequestA(Msg): ...
class ResponseA(Msg): ...
class ServiceA:
    Request: Final = RequestA
    Response: Final = ResponseA

def get_request_response[RequestT: Msg, ResponseT: Msg](
    x: Srv[RequestT, ResponseT],
) -> tuple[RequestT, ResponseT]: ...

v = get_request_response(ServiceA)
reveal_type(v)  # tuple[RequestA, ResponseA]
```

### Option 2: Describe the class not the object

Protocols are typically used to describe the shape of an object (a class instance), but they can be used to describe the shape of a class object (i.e. an instance of `type`). That may work for you in this case depending on whether there are other attributes and methods of `Srv` that you want to include in the protocol definition. This approach avoids the need to describe `Request` and `Response` as class variables. From the perspective of the class object, they are simply instance variables.

Here's the [relevant part](https://typing.readthedocs.io/en/latest/spec/protocol.html#assignability-relationships-with-other-types) of the typing spec. In particular, it says:

> A class object is considered an implementation of a protocol if accessing all members on it results in types assignable to the types of the protocol members.

Here's how you can apply this to your use case...

Code sample in [pyright playground](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoDCANgIYDOZAaiSADRQAK4MYAxmEQLABQPrpFKAFkyaABRMwLdkQCUALigA6FTz4CyUAMogAbsXJkA2gCUApgEcArmbIwAKopFp65sjhRkzj4aIC6EsxsHAo8UOFQ5ta2MIrwCGamljZ29n5hEW4eXnGIiVlgnt7pvNz8hpHJMQCCYs4KyqplGpXuhV619YoqSuoVWmZ6SKxm1fIZ4VEpsYQa1CBQALyV0XbVE63ZZooGFPNLm%2B2jatwAJmbAUGhmMAD6IFV297Zbt%2BUUtkmrDk6iri9HHzOAIbfhkRQ6fQaL7Tez-NpFNK0HiyKAAWgAfLArAgiIl4vlHg4-PQCUkEV40n5xtwIlAHjArCAUFAwUopjF6GyCkUTroDtc7g9vs8KWY3hpbGIBkMRtVZDwHrozCQiLcCWJdKioABibG4-F5GE1EmwI08jp%2BPxAA)

```python
from typing import ClassVar, Protocol

class Msg(Protocol): ...

class SrvClass[RequestT: Msg, ResponseT: Msg](Protocol):
    Request: type[RequestT]
    Response: type[ResponseT]

class RequestA(Msg): ...
class ResponseA(Msg): ...
class ServiceA:
    Request: ClassVar = RequestA
    Response: ClassVar = ResponseA

def get_request_response_classes[RequestT: Msg, ResponseT: Msg](
    cls: SrvClass[RequestT, ResponseT],
) -> tuple[type[RequestT], type[ResponseT]]:
    return cls.Request, cls.Response

v = get_request_response_classes(ServiceA)
reveal_type(v)  # tuple[type[RequestA], type[ResponseA]]
```
